developer_documentation/contribution_guidelines.Rmd

```diff
@@ -91,6 +92,86 @@ the devil].
  members of the group also have diverse backgrounds and likely can
  give valuable design insights.
 
+### Follow BioCro's git branching structure
+
+* In general, BioCro development follows the [git-flow][] branching model, where
+  there are two permanent branches (`main` and `develop`) and three types of
+  temporary branches (hotfixes, features, and releases).
+
+* For most contributors, it is only necessary to know that most changes should
+  be accomplished through feature branches, which are branched off `develop` and
+  merged back into `develop` via a pull request that must be approved by one or
+  more other developers. The remainder of this section discusses our system in
+  more detail.
+
+* Beyond the basic description of git-flow, we have a few additional rules
+  and clarifications specific to BioCro development:
+
+  - Any merge into `main` or `develop` must be done via a pull request. This
+    requirement is enforced with GitHub branch protection rules and cannot be
+    bypassed.
+
+  - All pull requests into `main` and `develop` require approval before merging.
+    This requirement is not enforced using GitHub branch protection rules, but
+    please do not unilaterally merge a pull request without any form of approval
+    from another developer. (There are two exceptions to this rule related to
+    hotfix and release branches, described below.)
+
+  - The BioCro R package and repository follow semantic versioning of the form
+    `major.minor.patch`. Hotfix branches should typically only increment the
+    third (`patch`) number. Most release branches will increment the second
+    (`minor`) number, and very rarely the first (`major`) number. Feature
+    branches do not directly change the version number.
+
+  - Hotfix and release branch names should be formatted like `type-version`,
+    where `type` is either `hotfix` or `release`, and version is the new version
+    number. For example, `hotfix-v3.0.3` would be a hotfix branch that
+    changes the version number to `3.0.3`. Feature branch names should reflect
+    their purpose, and can be anything other than `master`, `develop`,
+    `hotfix-*`, or `release-*`.
+
+  - `NEWS.md` should be updated whenever the package version changes, and
+    whenever a feature branch is merged into `develop`. In the latter case, a
+    description of the changes should be collected under an `UNRELEASED` section
+    in `NEWS.md` so it is easier to keep track of new changes. If there is no
+    `UNRELEASED` section in `NEWS.md` on `develop` when a feature branch is
+    merged into `develop`, one should be added.
```

Comment:
I think this section could be a little clearer, and here is my attempt at redrafting this:

- `NEWS.md` should be updated whenever the package version changes.  Normally, this will happen as part of the process of preparing a new release on a release branch:  A new heading specifying the version number of the new release is placed at the top of the content portion of the _news_ file `NEWS.md`.  This is followed by a listing of the important changes effected by the new release.  To make drafting the contents of this list easier, each time a developer works on a new feature branch, a new draft section labeled `UNRELEASED` should be placed at the top of the _news_ file, with the changes being made by the feature branch being listed underneath it.  Then, when it comes time to merge the feature branch into `develop`, the new `UNRELEASED` section is transferred into the `develop` branch's _news_ file, or, if the `develop` branch already has an `UNRELEASED` section in its _news_ file, the feature branch's `UNRELEASED` section will be integrated into the one on the `develop` branch.  (This process of integrating the two `UNRELEASED` sections will likely be part of resolving an inevitable merge conflict. )

    For details about the format of the _news_ file, see the comment at the top of `NEWS.md`.

    In the case of a hot fix, in addition to whatever substantive changes are made on the hotfix branch, a short section headed by the new release number should be added to the _news_ file to describe the needed fixes the hotfix branch addresses.

(The remark regarding the format of the _news_ file prompts me to suggest that the comment in `NEWS.md` should be revised.  Here's a possible updated version:

> This file should document user-visible changes brought about by each new release.
>
> To facilitate this, when a feature on a feature branch is completed and a pull request is being prepared, a new section should be added at the top of this file under the heading "# Unreleased"; it should list all the important changes made on the feature branch.   All changes should be categorized ...
>
> ... such as "# CHANGES IN BioCro VERSION 2.0.0." This section will combine the draft release notes for all features that have been added since the previous release.

(Omit the last sentence "Subsequent commits will then include a new "Unreleased" section in preparation
for the next release.")

)

Reply:
> I think this section could be a little clearer, and here is my attempt at redrafting this:
> 
> The remark regarding the format of the _news_ file prompts me to suggest that the comment in `NEWS.md` should be revised. Here's a possible updated version:

I generally like your suggestions. I tried implementing them with a few modifications.

I think the biggest change was to consolidate all the detailed instructions into the comment in `NEWS.md`, with a short overview in `contribution_guidelines.md`. To me, it seems convenient to have the instructions right there in `NEWS.md` when I start adding new changes.

Let me know what you think.